zapcore/sampler.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapcore
+
+import (
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap/internal/hash"
+)
+
+const (
+	_numLevels        = maxLevel - minLevel
+	_countersPerLevel = 4096
+)
+
+func newCounters() *counters {
+	return &counters{}
+}
+
+type counters [_numLevels][_countersPerLevel]uint64
```

Comment:
Well:
- I suspect that it will be more memory inefficient, if not actually slower, due to `atomic.Uint64` being a struct with a pointer field, rather than the actual `uint64` values living in this array
- the only use that we're getting from `go.uber.org/atomic` is for the `AtomicLevel` type, which is already a struct, and arguably would be simpler if it used standard `sync/atomic` functions directly

I'll benchmark to test any supposed difference, but more towards point 2, I really don't think that using `go.uber.org/atomic` is adding much value for zap at this point.